File: paper/irt.py
from numpy import exp
from ga import Chrom


class Item(object):
    """
    IRT模型M-2PPC中的一个被试项目
    具体模型定义参见bmirt
    """
    def __init__(self, index: int, data: list):
        """

        :param index: 项目编号
        :param data: 项目参数
        """
        self.index = index
        self.Kj = int(data[0])
        self.beta_2j = float(data[1])
        self.beta_delta_j = {}
        self.sum_beta = {}

        self.beta_delta_j[1] = 0.0
        for i in range(2, self.Kj + 1):
            self.beta_delta_j[i] = float(data[i])

        for i in range(1, max(self.beta_delta_j.keys()) + 1):
            self.sum_beta[i] = self.sum_beta_delta(i, self.beta_delta_j)

    def pk(self, theta: float) -> dict:
        d = {}

        for response in range(1, self.Kj + 1):
            d[response] = self.numerator(response, theta) / self.sum_denominator(self.Kj, theta)

        return d

    def EX(self, theta: float) -> float:
        dpk = self.pk(theta)
        sum = 0.0
        for k, v in dpk.items():
            sum += (k - 1) * v

        return sum

    def getParams(self) -> dict:
        d = {}

        d[1] = self.beta_2j
        for i in range(2, self.Kj + 1):
            d[i] = self.beta_delta_j[i]

        return d

    def sum_beta_delta(self, m: int, beta_delta_j: dict) -> float:
        sum = 0.0
        for i in range(1, m + 1):
            sum += beta_delta_j[i]

        return sum

    def denominator(self, m: int, theta: float) -> float:
        return exp((m - 1) * self.beta_2j * theta - self.sum_beta[m])

    def sum_denominator(self, K: int, theta: float) -> float:
        sum = 0.0
        for i in range(1, K + 1):
            sum += self.denominator(i, theta)

        return sum

    def numerator(self, k: int, theta: float) -> float:
        return exp((k - 1) * self.beta_2j * theta - self.sum_beta[k])

    def P_ijk(self, k: int, theta: float) -> float:
        numer = self.numerator(k, theta)
        # TODO 此处应该是sum_denominator
        denom = self.denominator(self.Kj, theta)

        return numer / denom


def delta(X_ij: int, k: int) -> bool:
    if X_ij == k:
        return True
    else:
        return False


def logP(N: int, J: int, KJ: dict, X_ij: dict, thetas: dict, params: dict, item: 'Chrom') -> float:
    sum = 0.0
    for i in range(1, N + 1):
        for j in range(1, J + 1):
            for k in range(1, KJ[j] + 1):
                allbeta_j = item.getAllBeta()[j]
                data = []
                data.append(KJ[j])
                data.append(allbeta_j[1])
                data.extend(list(allbeta_j.values()))
                tmp_item = Item(j, data)
                sum += tmp_item.P_ijk(k, thetas[i]) if delta(X_ij[i][j], k - 1) is True else 0

    return sum

File: paper_cs/paper/irt.cs
/*
 * 由SharpDevelop创建。
 * 用户： pfermat
 * 日期: 2017/12/14
 * 时间: 10:53
 * 
 * 要改变这种模板请点击 工具|选项|代码编写|编辑标准头文件
 */

using System;
using System.Collections.Generic;

public struct betaParam
{
	private double _2j;
	private double[] _delta_j;
	
	public betaParam( double beta_2j, double[] beta_delta_j )
	{
		_2j = beta_2j;
		_delta_j = new double[beta_delta_j.Length];
		for ( int i = 0; i < _delta_j.Length; i++ )
		{
			_delta_j[i] = beta_delta_j[i];
		}
	}
	
	public betaParam( double[] beta_data )
	{
		_2j = beta_data[0];
		_delta_j = new double[beta_data.Length - 1];
		for ( int i = 0; i < _delta_j.Length; i++ )
		{
			_delta_j[i] = beta_data[i + 1];
		}
	}
	
	/// <summary>
	/// 获取beta_2j的值
	/// </summary>
	/// <returns></returns>
	public double beta_2j( )
	{
		return _2j;
	}
	
	/// <summary>
	/// 直接返回beta_delta_j数组
	/// </summary>
	/// <remarks>下标从0开始</remarks>
	/// <returns></returns>
	public double[] beta_delta_j( )
	{
		return _delta_j;
	}
	
	/// <summary>
	/// 返回指定index的beta_delta_j
	/// </summary>
	/// <param name="index">下标</param>
	/// <remarks>下标从2开始</remarks>
	/// <returns></returns>
	public double beta_delta_j( int index )
	{
		if ( index == 0 || index == 1 )
		{
			return 0.0;
		}
		else
		{
			return _delta_j[index - 2];
		}
	}
	
	/// <summary>
	/// 返回所有beta参数
	/// </summary>
	/// <param name="type">1 下标1为0；2 下标1为beta_2j</param>
	/// <remarks>下标0无用，下标1为0.0，下标从2开始为beta_delta_j</remarks>
	/// <returns></returns>
	public double[] beta_data( int type = 1)
	{
		List<double> tmp = new List<double>();
		tmp.Add(0.0);
		if ( type == 1)
		{
			tmp.Add(0.0);
		}
		else
		{
			tmp.Add(_2j);
		}
		
		for ( int i = 0; i < _delta_j.Length; i++ )
		{
			tmp.Add(_delta_j[i]);
		}
		
		return tmp.ToArray();
	}
}

public struct Item
{
	public int Index;
	public int Kj;
	public betaParam Betas;
	public double[] SumBeta;

	public Item( int index, int kj, double[] beta_data )
	{
		Index = index;
		Kj = kj;
		Betas = new betaParam(beta_data);
		SumBeta = new double[Betas.beta_data().Length + 1];
		
		for ( int i = 1; i < Betas.beta_data().Length + 1; i++ )
		{
			SumBeta[i] = sum_beta_delta(i - 1);
		}
	}
	
	public Item( int index, int kj, betaParam beta_data )
	{
		Index = index;
		Kj = kj;
		Betas = beta_data;
		SumBeta = new double[Betas.beta_data().Length + 1];
		
		for ( int i = 1; i < Betas.beta_data().Length + 1; i++ )
		{
			SumBeta[i] = sum_beta_delta(i - 1);
		}
	}
	
	private double sum_beta_delta( int m )
	{
		double sum = 0.0;
		for ( int i = 1; i <= m; i++ )
		{
			sum += Betas.beta_data()[i];
		}
		
		return sum;
	}
	
	private double denominator( int m, double theta )
	{
		return Math.Exp( (m - 1) * Betas.beta_2j() * theta - SumBeta[m] );
	}
	
	private double numerator( int k, double theta )
	{
		return Math.Exp( (k - 1) * Betas.beta_2j() * theta - SumBeta[k] );
	}
	
	private double sum_denominator( int K, double theta )
	{
		double sum = 0.0;
		for ( int i = 1; i < K + 1; i++ )
		{
			sum += denominator(i, theta);
		}
		
		return sum;
	}
	
	public double P_ijk( int k, double theta )
	{
		double numer = numerator(k, theta);
		double denom = sum_denominator(Kj, theta);
		
		return numer / denom;
	}
	
	private double[] pk( double theta )
	{
		List<double> p = new List<double>();
		p.Add(0.0);
		
		for ( int i = 1; i < Kj + 1; i++ )
		{
			p[i] = numerator(i, theta) / sum_denominator(Kj, theta);
		}
		
		return p.ToArray();
	}
	
	public double EX( double theta )
	{
		double[] p = pk(theta);
		double sum = 0.0;
		
		for ( int i = 1; i < p.Length; i++ )
		{
			sum += (i - 1) * p[i];
		}
		
		return sum;
	}
}

public static class irt
{
	public static bool sig_delta( int X_ij, int k )
	{
		if ( X_ij == k )
		{
			return true;
		}
		else
		{
			return false;
		}
	}
	
	public static double logP( int N, int J, int[] KJ, int[,] X_ij, double[] thetas, betaParam[] betas )
	{
		double sum = 0.0;
		for ( int i = 0; i < N; i++ )
		{
			for ( int j = 0; j < J; j++ )
			{
				Item tmp = new Item(j, KJ[j], betas[j]);
				for ( int k = 0; k < KJ[j]; k++ )
				{
					if ( sig_delta(X_ij[i,j], k) )
					{
						sum += tmp.P_ijk(k + 1, thetas[i]);
					}
				}
			}
		}
		
		return sum;
	}
}